source/faq.txt

```diff
@@ -0,0 +1,169 @@
+   containing ``AreSessionsSupportedServerSelector``,
+   ``LatencyLimitingServerSelector``, and
+   ``OperationsCountServerSelector``).
+#. The driver’s current view of the cluster topology. The list of
+   servers that the driver is aware of is a key part of this view. Each
+   server description contains an exhaustive description of its current
+   state including information about an endpoint, a server version, a
+   server type, and its current health state. If the server is not
+   heathy, ``HeartbeatException`` will contain the exception from the
+   last failed heartbeat. Analyzing the ``HeartbeatException`` on each
+   cluster node can assist in diagnosing most server selection issues.
+   The following heartbeat exceptions are common:
+   
+   * ``No connection could be made because the target machine actively
+     refused it``: The driver cannot see this cluster node. This can be
+     because the cluster node has crashed, a firewall is preventing
+     network traffic from reaching the cluster node or port, or some other
+     network error is preventing traffic from being successfully routed to
+     the cluster node.
+   * ``Attempted to read past the end of the stream``: This error
+     happens when the driver cannot connect to the cluster nodes due to a
+     network error, misconfigured firewall, or other network issue. To
+     address this exception, ensure that all cluster nodes are reachable.
+     This error commonly occurs when the client machine’s IP address is
+     not configured in the Atlas IPs Access List, which can be found under
+     the Network Access tab for your Atlas Project.
+   * ``The remote certificate is invalid according to the validation
+     procedure``: This error typically indicates a TLS/SSL-related problem
+     such as an expired/invalid certificate or an untrusted root CA. You
+     can use tools like ``openssl s_client`` to debug TLS/SSL-related
+     certificate problems.
+
+Why is the Wait Queue for Acquiring a Connection to the Server Full?
+--------------------------------------------------------------------
+
+This exception usually indicates a threading or concurrency problem in
+your application. When the driver executes a read or write, it will
+check out a connection from the selected server’s connection pool. If
+that pool is already at ``maxPoolSize``, which is by default 100, then the requesting
+thread will block in a wait queue, which has a default size of ``5 x
+maxPoolSize`` or 500. If the wait queue is also full, then the driver
+throws a ``MongoWaitQueueFullException``. To resolve this issue, try the
+following steps:
+
+1. Tune your indexes. By improving the performance of your queries,
+   you can reduce the time that operations take and reduce the number of
+   concurrent connections needed for your workload.
+#. If you have long-running, analytical queries, you may wish to isolate
+   them to dedicated analytics nodes using ``readPreferenceTags`` or a hidden
```

Comment:
issue: it might be nice for users to be able to follow a link to read preference tag info.